README.md
# sts_flutter

AHCA STS Flutter plugin.
AHCA Flutter平台手机盾SDK插件

# 说明

此插件是在手机盾原生SDK上封装的Flutter平台插件，其功能和接口与原生SDK一致。
目前只支持Android平台的原生功能，iOS部分的待定。

## 初始化
初始化手机盾配置，建议在自定义Application类的onCreate生命方法内操作：
```
StsFlutter.init("服务地址", "AppKey", "SecretKey");
StsFlutter.initFace("活体检测授权码", "活体检测授权文件名");
```
注意，同时需要在确定用户信息和唯一标识后调用以下接口：
```
StsFlutter.initUseId("用户唯一标识");
```

## 可选的流程
请根据实际需求进行配置，例如项目无需使用手机盾活体检测功能，则无需配置initFace。
在使用证书前（例如签名、签章等），需要先判断证书是否已申请，如果证书不存在，需要调用申请证书接口，申请成功后方可使用。建议在调用证书签名/证书签章/证书登录等接口之前，先调用申请证书接口，当该接口返回码为1或10503时，再调用需要操作的证书功能的接口。
同时手机盾还提供了判断本地证书数据是否存在的接口，即isLocalCertExist接口，这是为了方便对流程响应速度有需求的业务场景，例如，APP证书登录场景，由于checkCert、certLogin或其他接口需要联网，存在一定的耗时，可仅使用isLocalCertExist接口仅判断本地证书是否存在，证书存在则调用证书登录接口，证书不存在则显示密码输入框或验证码相关控件，进行后续申请证书的操作。

## StsCodeTable
StsCodeTable类包含了SDK接口返回的部分错误码和对应解释，例如常用的 1(成功) 10502(参数不全或为空) 10502(证书不存在或已损坏) 10503(证书已存在)等

## StsFlutter
StsFlutter类包含接口外，还包括部分接口使用的固定标识
如申请个人证书接口的StsUserInfo实体类的证件类型属性(cardType)即可使用以下标识
```
  static const String CARD_TYPE_ID_CARD = "00"; // 身份证
  static const String CARD_TYPE_TEMPORARY_ID_CARD = "01"; // 临时身份证
  static const String CARD_TYPE_HOUSEHOLD_REGISTER = "02"; // 户口本
  static const String CARD_TYPE_PASSPORT = "03"; // 护照
  static const String CARD_TYPE_MILITARY_ID_CARD = "04"; // 军人身份证
  static const String CARD_TYPE_ARMED_POLICE_ID_CARD = "05"; // 武警身份证
```

如certSign、certLogin等接口的dataType、dataFormat参数可使用如下标识
```
  static const String DATA_TYPE_ORIGINAL_TO_HEXADECIMAL = "0"; // 原文数据转成16进制
  static const String DATA_TYPE_BYTE_BY_HEXADECIMAL = "1"; // byte类型数据（16进制编码）
  static const String DATA_TYPE_CHINESE_TO_HEXADECIMAL = "2"; // 中文转成16进制
  static const String DATA_TYPE_ORIGINAL = "3"; // 原文数据
  static const String DATA_TYPE_BYTE_BY_BASE64 = "4"; // byte类型数据（Base64编码加密后的值）

  static const String DATA_FORMAT_P1 = "p1"; // P1
  static const String DATA_FORMAT_P7 = "p7"; // P7
  static const String DATA_FORMAT_P7_ENVELOPE = "p7envelope"; // P7数字信封
  static const String DATA_FORMAT_DEFAULT = "default"; // 裸签名
```

CHANGELOG.md


## 0.0.18
* 更新STS iOS SDK，修复release模式的部分问题。

## 0.0.17
* 修改示例部分，增加企业证书分支；
* 修改iOS示例bundleId，修改跨平台接口中的企业证书相关接口；
* 修改Android示例包名，新增人脸识别测试流程；

## 0.0.16
* 更新Android端手机盾SDK；

## 0.0.15
* 修改podspec配置；
* 修改录入用户信息页数据初始化；

## 0.0.14
* 更新Android原生SDK；
* 更新iOS原生SDK；
* 修改iOS目录资源结构，修改跨平台通道代码；
* 修改启动页初始化，修改录入用户信息页；

## 0.0.13
* 修改iOS平台配置以及代码，修复已知错误，消除警告；

## 0.0.12
* 新增iOS平台部分功能；

## 0.0.11
* 更新Android手机盾SDK；

## 0.0.10
* 更新Android手机盾SDK和调用接口；

## 0.0.9
* 更新gradle和gradle插件的版本；

## 0.0.8
* 更新原生SDK，修改平台接口传参类型；

## 0.0.7
* 修改平台接口传参类型，修复部分接口无法调用的问题；

## 0.0.6
* 更新手机盾Android平台原生SDK为 2.8.1；

## 0.0.5
* 更新手机盾Android平台原生SDK，修改接口调用；

## 0.0.4
* 修改Android原生的签名；

## 0.0.3
* 修改了接口类的类名和发布说明；

## 0.0.2
* 修改了接口类的类名；
* 修改了发布说明；
* 修改主页地址；

## 0.0.1
* TODO: Describe initial release.



lib/sts_flutter.dart
import 'dart:convert';

import 'package:sts_flutter/model/sts_user_info.dart';
import 'package:sts_flutter/model/sts_company_info.dart';
import 'package:sts_flutter/model/sts_scan_info.dart';
import 'package:sts_flutter/model/sts_sign_img_setting.dart';
import 'package:sts_flutter/result/apply_cert_result.dart';
import 'package:sts_flutter/result/cert_decrypt_result.dart';
import 'package:sts_flutter/result/cert_encrypt_result.dart';
import 'package:sts_flutter/result/cert_login_result.dart';
import 'package:sts_flutter/result/cert_seal_result.dart';
import 'package:sts_flutter/result/cert_sign_result.dart';
import 'package:sts_flutter/result/common_result.dart';
import 'package:sts_flutter/result/get_cert_result.dart';
import 'package:sts_flutter/result/get_department_no_result.dart';
import 'package:sts_flutter/result/get_qr_code_result.dart';
import 'package:sts_flutter/result/pk_cache_result.dart';
import 'package:sts_flutter/result/sign_img_result.dart';
import 'package:sts_flutter/sts_code_table.dart';
import 'package:sts_flutter/sts_platform_bridge.dart';

class StsFlutter {
  /// [.applyPersonalCert]
  /// [.applyPersonalCertWithPIN]
  /// [.updatePersonalCert]
  /// [.resetPersonalPIN]
  static const String CARD_TYPE_ID_CARD = "00"; // 身份证
  static const String CARD_TYPE_TEMPORARY_ID_CARD = "01"; // 临时身份证
  static const String CARD_TYPE_HOUSEHOLD_REGISTER = "02"; // 户口本
  static const String CARD_TYPE_PASSPORT = "03"; // 护照
  static const String CARD_TYPE_MILITARY_ID_CARD = "04"; // 军人身份证
  static const String CARD_TYPE_ARMED_POLICE_ID_CARD = "05"; // 武警身份证

  static const String CARD_TYPE_ID_CARD_NAME = "身份证"; // 身份证
  static const String CARD_TYPE_TEMPORARY_ID_CARD_NAME = "临时身份证"; // 临时身份证
  static const String CARD_TYPE_HOUSEHOLD_REGISTER_NAME = "户口本"; // 户口本
  static const String CARD_TYPE_PASSPORT_NAME = "护照"; // 护照
  static const String CARD_TYPE_MILITARY_ID_CARD_NAME = "军人身份证"; // 军人身份证
  static const String CARD_TYPE_ARMED_POLICE_ID_CARD_NAME = "武警身份证"; // 武警身份证

  /// [.certLogin]
  /// [.certSign]
  /// [.certVerifySign]
  /// [.certEncrypt]
  /// [.certDecrypt]
  static const String DATA_TYPE_ORIGINAL_TO_HEXADECIMAL = "0"; // 原文数据转成16进制
  static const String DATA_TYPE_BYTE_BY_HEXADECIMAL = "1"; // byte类型数据（16进制编码）
  static const String DATA_TYPE_CHINESE_TO_HEXADECIMAL = "2"; // 中文转成16进制
  static const String DATA_TYPE_ORIGINAL = "3"; // 原文数据
  static const String DATA_TYPE_BYTE_BY_BASE64 = "4"; // byte类型数据（Base64编码加密后的值）

  static const String DATA_FORMAT_P1 = "p1"; // P1
  static const String DATA_FORMAT_P7 = "p7"; // P7
  static const String DATA_FORMAT_P7_ENVELOPE = "p7envelope"; // P7数字信封
  static const String DATA_FORMAT_DEFAULT = "default"; // 裸签名

  /// [.changeCertStatus]
  static const int CERT_STATUS_TYPE_FREEZE = 1; // 冻结证书
  static const int CERT_STATUS_TYPE_UNFREEZE = 2; // 解冻证书
  static const int CERT_STATUS_TYPE_REVOKE = 3; // 销毁证书

  /// [.getCert]
  static const int CERT_TYPE_SIGNCERT = 1; // 签名证书
  static const int CERT_TYPE_ENCCERT = 2; // 解密证书

  /// [.setSignImgWithDrawingBoard]
  /// [.getSignImgAndSetItIfNotExist]
  /// [.getSignImgWithDrawingBoard]
  static const int ORIENTATION_AUTO = 1; // 根据系统切换
  static const int ORIENTATION_LANDSCAPE = 2; // 横屏
  static const int ORIENTATION_PORTRAIT = 3; // 竖屏

//  证书相关
  static Future<String> init(String url, String appKey, String secretKey) async {
    return await StsPlatformBridge.init(url, appKey, secretKey);
  }

  static Future<String> initUseId(String useId) async {
    return await StsPlatformBridge.initUseId(useId);
  }

  static Future<String> initFace(String id, String fileName) async {
    return await StsPlatformBridge.initFace(id, fileName);
  }

  static Future<String> initThemeColor(int color) async {
    return await StsPlatformBridge.initThemeColor(color);
  }

  static Future<String> get getDeviceID async {
    return await StsPlatformBridge.getDeviceID;
  }

  static Future<bool> get isLocalCertExist async {
    return await StsPlatformBridge.isLocalCertExist;
  }

  static Future<CommonResult> get checkCert async {
    String jsonStr = await StsPlatformBridge.checkCert;
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> get clearCert async {
    String jsonStr = await StsPlatformBridge.clearCert;
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<ApplyCertResult> applyPersonalCert(StsUserInfo stsUserInfo) async {
    if (stsUserInfo == null)
      return ApplyCertResult(
          resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data, enCert: null, signCert: null, stsCertInfo: null, stsUserInfo: null, stsCompanyInfo: null);
    String jsonStr = await StsPlatformBridge.applyPersonalCert(json.encode(stsUserInfo));
    return ApplyCertResult.fromJson(json.decode(jsonStr));
  }

  static Future<ApplyCertResult> applyCompanyCert(StsCompanyInfo stsCompanyInfo) async {
    if (stsCompanyInfo == null)
      return ApplyCertResult(
          resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data, enCert: null, signCert: null, stsCertInfo: null, stsUserInfo: null, stsCompanyInfo: null);
    String jsonStr = await StsPlatformBridge.applyCompanyCert(json.encode(stsCompanyInfo));
    return ApplyCertResult.fromJson(json.decode(jsonStr));
  }

  static Future<GetQRCodeResult> get getUntieEquipmentQRCode async {
    String jsonStr = await StsPlatformBridge.getUntieEquipmentQRCode;
    return GetQRCodeResult.fromJson(json.decode(jsonStr));
  }

  static Future<ApplyCertResult> untieEquipment(String qrData) async {
    String jsonStr = await StsPlatformBridge.untieEquipment(qrData);
    return ApplyCertResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> updatePersonalCert(StsUserInfo stsUserInfo) async {
    if (stsUserInfo == null) return CommonResult(resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data);
    String jsonStr = await StsPlatformBridge.updatePersonalCert(json.encode(stsUserInfo));
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> updateCompanyCert(StsCompanyInfo stsCompanyInfo) async {
    if (stsCompanyInfo == null) return CommonResult(resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data);
    String jsonStr = await StsPlatformBridge.updateCompanyCert(json.encode(stsCompanyInfo));
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> resetPersonalPIN(StsUserInfo stsUserInfo) async {
    if (stsUserInfo == null) return CommonResult(resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data);
    String jsonStr = await StsPlatformBridge.resetPersonalPIN(json.encode(stsUserInfo));
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> resetCompanyPIN(StsCompanyInfo stsCompanyInfo) async {
    if (stsCompanyInfo == null) return CommonResult(resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data);
    String jsonStr = await StsPlatformBridge.resetCompanyPIN(json.encode(stsCompanyInfo));
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> get modifyPIN async {
    String jsonStr = await StsPlatformBridge.modifyPIN;
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> scanLogin(StsScanInfo stsScanInfo) async {
    if (stsScanInfo == null) return CommonResult(resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data);
    String jsonStr = await StsPlatformBridge.scanLogin(json.encode(stsScanInfo));
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> scanSign(StsScanInfo stsScanInfo) async {
    if (stsScanInfo == null) return CommonResult(resultCode: StsCodeTable.rtnCode_no_data, resultMsg: StsCodeTable.rtnMsg_no_data);
    String jsonStr = await StsPlatformBridge.scanSign(json.encode(stsScanInfo));
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CertLoginResult> certLogin(String data, String dataFormat, String dataType, String pn) async {
    String jsonStr = await StsPlatformBridge.certLogin(data, dataFormat, dataType, pn);
    return CertLoginResult.fromJson(json.decode(jsonStr));
  }

  static Future<CertSealResult> certSeal(String pn) async {
    String jsonStr = await StsPlatformBridge.certSeal(pn);
    return CertSealResult.fromJson(json.decode(jsonStr));
  }

  static Future<CertSignResult> certSign(String data, String dataFormat, String dataType, String pn) async {
    String jsonStr = await StsPlatformBridge.certSign(data, dataFormat, dataType, pn);
    return CertSignResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> certVerifySign(String data, String dataFormat, String signData, String dataType) async {
    String jsonStr = await StsPlatformBridge.certVerifySign(data, dataFormat, signData, dataType);
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CertEncryptResult> certEncrypt(String data, String dataFormat, String dataType) async {
    String jsonStr = await StsPlatformBridge.certEncrypt(data, dataFormat, dataType);
    return CertEncryptResult.fromJson(json.decode(jsonStr));
  }

  static Future<CertDecryptResult> certDecrypt(String data, String dataFormat, String dataType, String pn) async {
    String jsonStr = await StsPlatformBridge.certDecrypt(data, dataFormat, dataType, pn);
    return CertDecryptResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> changeCertStatus(int statusType) async {
    String jsonStr = await StsPlatformBridge.changeCertStatus(statusType);
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<CommonResult> get postponeCert async {
    String jsonStr = await StsPlatformBridge.postponeCert;
    return CommonResult.fromJson(json.decode(jsonStr));
  }

  static Future<PKCacheResult> setPKCacheTime(String pn) async {
    String jsonStr = await StsPlatformBridge.setPKCacheTime(pn);
    return PKCacheResult.fromJson(json.decode(jsonStr));
  }

  static Future<PKCacheResult> clearPKCacheTime(String pn) async {
    String jsonStr = await StsPlatformBridge.clearPKCacheTime(pn);
    return PKCacheResult.fromJson(json.decode(jsonStr));
  }

  static Future<GetCertResult> getCert(int certType) async {
    String jsonStr = await StsPlatformBridge.getCert(certType);
    return GetCertResult.fromJson(json.decode(jsonStr));
  }

  static Future<GetCertResult> getLocalCert(int certType) async {
    String jsonStr = await StsPlatformBridge.getLocalCert(certType);
    return GetCertResult.fromJson(json.decode(jsonStr));
  }

  static Future<GetCertResult> downloadCert(String phoneNum, String departmentNo, int certType) async {
    String jsonStr = await StsPlatformBridge.downloadCert(phoneNum, departmentNo, certType);
    return GetCertResult.fromJson(json.decode(jsonStr));
  }

  static Future<GetDepartmentNoResult> get getDepartmentNo async {
    String jsonStr = await StsPlatformBridge.getDepartmentNo;
    return GetDepartmentNoResult.fromJson(json.decode(jsonStr));
  }

  static Future<SignImgResult> setSignImgWithDrawingBoard(StsSignImgSetting stsSignImgSetting) async {
    String jsonStr = await StsPlatformBridge.setSignImgWithDrawingBoard(stsSignImgSetting == null ? null : json.encode(stsSignImgSetting));
    return SignImgResult.fromJson(json.decode(jsonStr));
  }

  static Future<SignImgResult> setSignImgBase64Str(String imgBase64) async {
    String jsonStr = await StsPlatformBridge.setSignImgBase64Str(imgBase64);
    return SignImgResult.fromJson(json.decode(jsonStr));
  }

  static Future<SignImgResult> get getSignImgFromService async {
    String jsonStr = await StsPlatformBridge.getSignImgFromService;
    return SignImgResult.fromJson(json.decode(jsonStr));
  }

  static Future<SignImgResult> getSignImgAndSetItIfNotExist(StsSignImgSetting stsSignImgSetting) async {
    String jsonStr = await StsPlatformBridge.getSignImgAndSetItIfNotExist(stsSignImgSetting == null ? null : json.encode(stsSignImgSetting));
    return SignImgResult.fromJson(json.decode(jsonStr));
  }

  static Future<SignImgResult> getSignImgWithDrawingBoard(StsSignImgSetting stsSignImgSetting) async {
    String jsonStr = await StsPlatformBridge.getSignImgWithDrawingBoard(stsSignImgSetting == null ? null : json.encode(stsSignImgSetting));
    return SignImgResult.fromJson(json.decode(jsonStr));
  }

  static Future<bool> get getFingerprintStatus {
    return StsPlatformBridge.getFingerprintStatus;
  }

  static Future<CommonResult> openFingerprint(bool open) async {
    String jsonStr = await StsPlatformBridge.openFingerprint(open);
    return CommonResult.fromJson(json.decode(jsonStr));
  }
}
